geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java

```diff
@@ -154,207 +165,274 @@
-  FLUSHDB(new FlushAllExecutor(), UNSUPPORTED, new MaximumParameterRequirements(2, ERROR_SYNTAX)),
-  SHUTDOWN(new ShutDownExecutor(), UNSUPPORTED, new MaximumParameterRequirements(2, ERROR_SYNTAX)),
-  TIME(new TimeExecutor(), UNSUPPORTED, new ExactParameterRequirements(1)),
+  DBSIZE(new DBSizeExecutor(), UNSUPPORTED, new Parameter().exact(1).firstKey(0).flags(readonly, fast)),
+  FLUSHALL(new FlushAllExecutor(), UNSUPPORTED,
+      new Parameter().min(1).max(2, ERROR_SYNTAX).firstKey(0).flags(write)),
+  FLUSHDB(new FlushAllExecutor(), UNSUPPORTED,
+      new Parameter().min(1).max(2, ERROR_SYNTAX).firstKey(0).flags(write)),
+  SHUTDOWN(new ShutDownExecutor(), UNSUPPORTED,
+      new Parameter().min(1).max(2, ERROR_SYNTAX).firstKey(0).flags(admin, noscript, loading,
+          stale)),
+  TIME(new TimeExecutor(), UNSUPPORTED,
+      new Parameter().exact(1).firstKey(0).flags(random, loading, stale, fast)),
 
   /***************************************
    *********** Unknown Commands **********
    ***************************************/
   UNKNOWN(new UnknownExecutor(), RedisCommandSupportLevel.UNKNOWN);
 
+  public enum Flag {
+    admin,
+    denyoom,
+    fast,
+    loading,
+    may_replicate,
+    movablekeys,
+    no_auth,
+    noscript,
+    pubsub,
+    random,
+    readonly,
+    sort_for_script,
+    stale,
+    write;
+  }
+
   private final Executor executor;
-  private final ParameterRequirements parameterRequirements;
-  private final ParameterRequirements deferredParameterRequirements;
+  private final Parameter parameterRequirements;
+  private final Parameter deferredParameterRequirements;
   private final RedisCommandSupportLevel supportLevel;
 
   RedisCommandType(Executor executor, RedisCommandSupportLevel supportLevel) {
-    this(executor, supportLevel, new UnspecifiedParameterRequirements());
+    this(executor, supportLevel, new Parameter().custom((c, e) -> {
+    }));
   }
 
   RedisCommandType(Executor executor, RedisCommandSupportLevel supportLevel,
-      ParameterRequirements parameterRequirements) {
-    this(executor, supportLevel, parameterRequirements, new UnspecifiedParameterRequirements());
+      Parameter parameterRequirements) {
+    this(executor, supportLevel, parameterRequirements, new Parameter().custom((c, e) -> {
+    }));
   }
 
   RedisCommandType(Executor executor, RedisCommandSupportLevel supportLevel,
-      ParameterRequirements parameterRequirements,
-      ParameterRequirements deferredParameterRequirements) {
+      Parameter parameterRequirements,
+      Parameter deferredParameterRequirements) {
     this.executor = executor;
     this.supportLevel = supportLevel;
     this.parameterRequirements = parameterRequirements;
     this.deferredParameterRequirements = deferredParameterRequirements;
   }
 
+  public int arity() {
+    return parameterRequirements.getArity();
+  }
+
+  public List<Flag> flags() {
```

Comment:
instead of a List<Flag> should this be EnumSet<Flag>?

Reply:
Cool - haven't used that before.